main/src/main/scala/sbt/Defaults.scala

```diff
@@ -1625,6 +1634,28 @@ object Defaults extends BuildCommon {
           else old
         }
       ))
+
+  private case class JavaDiscoverConf(root: String, dirRexp: Regex, innerDir: String)
+  private final val JavaDiscoverConfigs = Seq(
+    JavaDiscoverConf("/usr/lib/jvm", "java-([0-9]+)-.*".r, ""), // most linux
+    JavaDiscoverConf("/Library/Java/JavaVirtualMachines",
+                     "jdk[1\\.]*([0-9]+)\\..*".r,
+                     "Contents/Home") // mac os
+  )
+
+  private[sbt] def discoverJavaHomes: Initialize[Map[String, File]] = Def.setting {
+    import scala.collection.breakOut
+    JavaDiscoverConfigs
+      .flatMap {
+        case JavaDiscoverConf(root, dirRexp, inner) =>
+          Option(new File(root).list())
+            .getOrElse(Array.empty[String])
+            .toSeq
+            .collect {
+              case dir @ dirRexp(ver) => ver -> Paths.get(root, dir, inner).toFile
+            }
+      }(breakOut)
```

Comment:
With `breakOut` making an exit in Scala 2.13, best implement this with `.iterator` (after `JavaDiscoverConfigs`) and a final `.toMap`